src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -826,52 +828,93 @@ public void ApplyLogsToDataFile(CancellationToken token, TimeSpan timeToWait)
-                    _journalsToDelete.Values.ToList()));
+                var oldestActiveTransaction = _waj._env.ActiveTransactions.OldestTransaction;
 
-                if (unusedJournals.Count > 0)
+                IDisposable exitPreventNewTransactions = null;
+
+                try
                 {
-                    var lastUnusedJournalNumber = unusedJournals[unusedJournals.Count - 1].Number;
-                    _waj._files = _waj._files.RemoveWhile(x => x.Number <= lastUnusedJournalNumber);
-                }
+                    // we release up to the last read transaction, because there might be new read transactions that are currently
+                    // running, that started after the flush
+                    var lastFlushedTransactionIdThatWontReadFromJournal = Math.Min(Math.Min(lastFlushedTransactionId, _waj._env.CurrentReadTransactionId - 1), txw.Id - 1);
 
-                if (_waj._files.Count == 0)
-                    _waj.CurrentFile = null;
+                    if (oldestActiveTransaction == txw.Id && journalSnapshots.Count > 0 && journalSnapshots[^1].Number == lastProcessedJournal)
+                    {
+                        // we're the only active transaction and there are no reading transactions
+                        // let's try to prevent running new read transactions so we'll be able to free more scratch pages and delete more journals
+                        // in particular we might clean the last journal so we won't need to apply it on restart - see RavenDB-11871 
+
+                        if (_waj._env.IsInPreventNewReadTransactionsMode || _waj._env.TryPreventNewReadTransactions(TimeSpan.Zero, out exitPreventNewTransactions))
+                        {
+                            // we managed to prevent from new read transactions, let's verify that we're still the only active transaction
 
-                // we release up to the last read transaction, because there might be new read transactions that are currently
-                // running, that started after the flush
-                var lastSyncedTransactionId =
-                    Math.Min(Math.Min(lastFlushedTransactionId, _waj._env.CurrentReadTransactionId - 1), txw.Id - 1);
+                            var oldestActiveTransactionAfterPreventingNewReadTransactions = _waj._env.ActiveTransactions.OldestTransaction;
 
-                // we have to free pages of the unused journals before the remaining ones that are still in use
-                // to prevent reading from them by any read transaction (read transactions search journals from the newest
-                // to read the most updated version)
+                            if (oldestActiveTransactionAfterPreventingNewReadTransactions == txw.Id)
+                            {
+                                lastFlushedTransactionIdThatWontReadFromJournal = Math.Min(lastFlushedTransactionId, txw.Id - 1);
```

Comment:
We're doing this specifically to delete the last journal (potentially large) in order to avoid applying it after the restart. See https://github.com/ravendb/ravendb/pull/8326 for reference.